### ISSUE TITLE
Various features missing for variable used as argument function-like macro call

### ISSUE THREAD

**Issue by ckandeler** [clangd/clangd#1844]
Consider this simple piece of code:
```
#define MY_MACRO(param)

int main()
{
    int arg = 0;
    MY_MACRO(arg);
}

```
With the cursor on the second occurrence of the "arg" variable, at least the following features are not available:

- document highlights
- semantic highlighting
- find references

The first one appears to be due to no entries in the selection tree, the other ones I have not debugged into.

Output of `clangd --version`:
clangd version 18.0.0 (https://github.com/llvm/llvm-project.git d1652ff0803ac9f2f3ea99336f71edacdf95a721)
Features: linux+debug
Platform: x86_64-unknown-linux-gnu

**Comment by ckandeler** [comment]
It all works if the macro uses the parameter, e.g.:
#define MY_MACRO(param) ((void) param)

**Comment by ckandeler** [comment]
So, just another instance of https://github.com/clangd/clangd/issues/1159.

**Comment by HighCommander4** [comment]
It's related to #1159, though the fix discussed in #1159 would be specific to code completion, and other editor features would need their own fixes.

To illustrate the case of semantic highlighting as an example, consider:

```c++
namespace N {
  int arg;
}

struct S {
  static int arg;
};

#define MACRO1(param) N::param
#define MACRO2(param) S::param
#define MACRO3(param) param

int main() {
  int arg;
  MACRO1(arg);  // semantic token is variable.globalScope
  MACRO2(arg);  // semantic token is variable.classScope
  MACRO3(arg);  // senantic token is variable.functionScope
}
```

So, I guess the suggested behavior, if a macro does not use one of its parameters, treat that argument as though it was written in the context of the macro invocation (e.g. here, at local scope, so it would behave as `MACRO3`)?

Is this something that comes up a lot / would be a useful heuristic in your experience?

**Comment by ckandeler** [comment]
> int main() {
>   int arg;
>   MACRO1(arg);  // semantic token is variable.globalScope
>   MACRO2(arg);  // semantic token is variable.classScope
>   MACRO3(arg);  // semantic token is variable.functionScope
> }

My initial (naive?) expectation as a user would be that the expansion is completely irrelevant here -- after all, "arg" is a local variable in all three cases. For analogy, consider the renaming case: Clearly, when the cursor is on any of the local "arg" occurrences, you always want to rename the local variable and never the struct member.

> So, I guess the suggested behavior is, if a macro does not use one of its parameters, treat that argument as though 
> it was written in the context of the macro invocation (e.g. here, at local scope, so it would behave as `MACRO3`)?

Sounds sensible to me.

> Is this something that comes up a lot / would be a useful heuristic in your experience?

Well, I think this is a popular pattern:

```
#ifdef SOME_FEATURE
#define SOME_MACRO(x) builtin_feature(x)
#else
#define SOME_MACRO(x) 
#endif

```
So I think this issue should occur not so rarely.

**Comment by HighCommander4** [comment]
> > int main() {
> > int arg;
> > MACRO1(arg);  // semantic token is variable.globalScope
> > MACRO2(arg);  // semantic token is variable.classScope
> > MACRO3(arg);  // semantic token is variable.functionScope
> > }
> 
> My initial (naive?) expectation as a user would be that the expansion is completely irrelevant here -- after all, "arg" is a local variable in all three cases. 

Right, this would be another way to model the invocations of function-like macros: as if the macro was an actual function, with the macro expansion not affecting the analysis of the call much like the contents of a function's body doesn't affect the analysis of a function call.

The reason we don't currently try to model macros this way is that clangd is based on the internals of the clang compiler, and most features are implemented in terms of the compiler's AST, and in the AST macro invocations have been replaced by their expansions (but function calls have not been replaced but their bodies).

So trying to model things this way would definitely go against the grain of how clangd currently works. But from a conceptual point of view, it's possible.

> For analogy, consider the renaming case: Clearly, when the cursor is on any of the local "arg" occurrences, you always want to rename the local variable and never the struct member.

I'm not sure about that. Some of these invocations are valid without a local variable existing:

```c++
namespace N {
  int arg;
}

#define MACRO1(param) N::param

int main() {
  MACRO1(arg);
}
```

Now there is nothing to rename. (I'm assuming we still wouldn't want to rename the namespace-scope variable.)

Now suppose we add back the local variable:

```c++
namespace N {
  int arg;
}

#define MACRO1(param) N::param

int main() {
  int arg;
  MACRO1(arg);
}
```

It has no impact on the meaning of the `MACRO1(arg)` call, so it would be strange if now we would start to treat the `arg` in the macro invocation as an occurrence of it.

> Well, I think this is a popular pattern:
> 
> ```
> #ifdef SOME_FEATURE
> #define SOME_MACRO(x) builtin_feature(x)
> #else
> #define SOME_MACRO(x) 
> #endif
> ```
> 
> So I think this issue should occur not so rarely.

Right, that's fair, I've run into cases like this as well.

One challenge here is that the argument is not necessarily an identifier, it could be a complex expression.

Cosnider this example:

```c++
assert(is_sorted(widgets, [](const widget& a, const widget& b){ return a.frob > b.frob; }));
```

Here, the macro argument is a complex expression. From the point of view of semantic coloring, for example, it contains a mixture of local variables, types, parameters, and fields.

Clangd assigns these tokens based on the AST, but if the macro expansion is empty, an AST is never built for this expression.

So, I'm not quite sure what we'd do here:
 * Try to build an AST for the expression anyways?
    * This could have undesirable side effects such as diagnostics (e.g. imagine the function `is_sorted` itself being defined inside an `#ifndef NDEBUG` block).
 * Try to assign semantic tokens to parts of the expression heuristically, without an AST?
    * This sort of heuristic parsing would be new territory for clangd. In VSCode, we already have a heuristic (regex-based) coloring engine on the client side; I'm not sure that we could do much better on the server side without an AST.
